pkg/arvo/sys/hoon.hoon

```diff
@@ -1897,6 +1898,1491 @@
+  ::
+  +$  buc  [=k =v t=pro]                                ::  bucket
+  ::
+  +$  pri                                               ::  psq
+    $@  ~
+    $%  [%bin k=@ p=@ v=buc m=@ l=pri r=pri]
+        [%tip k=@ p=@ v=buc]
+    ==
+  --
+  =>
+  ::  internals
+  ::
+  ~%  %in  ..$  ~
+  |%
+  ::  balancing, tournament internals
+  ::
+  ++  mega  4                                           ::  balancing factor
+  ::
+  ++  size                                              ::  queue size
+    ~/  %size
+    |=  t=ltree
+    ^-  @
+    ?~(t 0 s.t)
+  ::
+  ++  one                                               ::  construct
+    ~/  %one
+    |=  =elem
+    ^-  pro
+    [elem ~ k.elem]
+  ::
+  ++  llos                                              ::  left loser
+    ~/  %llos
+    |=  a=lnode
+    ^-  ltree
+    [%llos +((add (size l.a) (size r.a))) a]
+  ::
+  ++  rlos                                              ::  right loser
+    ~/  %rlos
+    |=  a=lnode
+    ^-  ltree
+    [%rlos +((add (size l.a) (size r.a))) a]
+  ::
+  ++  lbal                                              ::  left balance
+    ~/  %lbal
+    |=  a=lnode
+    ^-  ltree
+    ?:  (lth (add (size l.a) (size r.a)) 2)
+      (llos a)
+    ?:  (gth (size r.a) (mul mega (size l.a)))
+      (llbal a)
+    ?:  (gth (size l.a) (mul mega (size r.a)))
+      (lrbal a)
+    (llos a)
+  ::
+  ++  rbal                                              ::  right balance
+    ~/  %rbal
+    |=  a=lnode
+    ^-  ltree
+    ?:  (lth (add (size l.a) (size r.a)) 2)
+      (rlos a)
+    ?:  (gth (size r.a) (mul mega (size l.a)))
+      (rlbal a)
+    ?:  (gth (size l.a) (mul mega (size r.a)))
+      (rrbal a)
+    (rlos a)
+  ::
+  ++  llbal                                             ::  left balance left
+    ~/  %llbal
+    |=  a=lnode
+    ^-  ltree
+    ?>  ?=(^ r.a)
+    ?:  (lth (size l.p.r.a) (size r.p.r.a))
+      (llsin a)
+    (lldub a)
+  ::
+  ++  lrbal                                             ::  left balance right
+    ~/  %lrbal
+    |=  a=lnode
+    ^-  ltree
+    ?>  ?=(^ l.a)
+    ?:  (gth (size l.p.l.a) (size r.p.l.a))
+      (lrsin a)
+    (lrdub a)
+  ::
+  ++  rlbal
+    ~/  %rlbal
+    |=  a=lnode
+    ^-  ltree
+    ?>  ?=(^ r.a)
+    ?:  (lth (size l.p.r.a) (size r.p.r.a))
+      (rlsin a)
+    (rldub a)
+  ::
+  ++  rrbal
+    ~/  %rrbal
+    |=  a=lnode
+    ^-  ltree
+    ?>  ?=(^ l.a)
+    ?:  (gth (size l.p.l.a) (size r.p.l.a))
+      (rrsin a)
+    (rrdub a)
+  ::
+  ++  llsin                                             ::  left single left
+    ~/  %llsin
+    |=  a=lnode
+    ^-  ltree
+    ?>  ?=(^ r.a)
+    ?-  -.r.a
+        %llos
+      ?:  (win [p.n.a k.n.a] [p.n.p.r.a k.n.p.r.a])
+        (llos n.a (rlos n.p.r.a l.a m.a l.p.r.a) m.p.r.a r.p.r.a)
+      (llos n.p.r.a (llos n.a l.a m.a l.p.r.a) m.p.r.a r.p.r.a)
+        %rlos
+      (rlos n.p.r.a (llos n.a l.a m.a l.p.r.a) m.p.r.a r.p.r.a)
```

Comment:
I hate to say it, but I find this almost completely unreadable, along with a bunch of the following functions.

Reply:
Yes, @tacryt-socryp mentioned the same.  My rationale for the terseness here is that, as data structure internals, these are sort of bound to be a little inscrutable anyway.  But given there are now two "this could be clearer" comments, I'll look at cleaning them up a little.